Shortest path query

ashwin95r opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph/570)
<!-- Reviewable:end -->

gacharya255 commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[query/shortest.go, line 16 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcPune1NDMOTSf5CCFP:-KcPune1NDMOTSf5CCFQ:b-3h3r1s) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L16)):*
> ```Go
> 	cost  float64 // cost of taking the path till this uid.
> 	hop   int     // number of hops taken to reach this node.
> 	index int
> ```

index tells you when it is added but don't think you really need that?

---

*[query/shortest.go, line 92 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcPwT6_zGWATDa7Dn0_:-KcPwT6_zGWATDa7Dn0a:b-o2pnus) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L92)):*
> ```Go
> 		var out []*SubGraph
> 		for _, sg := range exec {
> 			for _, child := range start.Children {
> ```

Checking my understanding: The expansion is only along start.Children which is why you need to call ProcessGraph again and again here.

---

*[query/shortest.go, line 157 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcQ4ndR3p1So4MG7enV:-KcQ4ndR3p1So4MG7enW:bxst53b) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L157)):*
> ```Go
> 					break
> 				}
> 				node := &Item{it, item.cost + 1, item.hop + 1, 0}
> ```

I thought you have cost because in the future, it might be a variable. However, in that future case, the code may not work? Two hops may lead to lower cost than one hop, and you would need to update a node in the priority queue and move it around?

---

*[query/shortest.go, line 160 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcQ3zHpyhFcoxP_6ETd:-KcQ3zHpyhFcoxP_6ETe:b-8n9y5k) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L160)):*
> ```Go
> 				heap.Push(&pq, node)
> 			}
> 			numHops++
> ```

The first time you see a node at a fixed distance (num hops), you find its neighbors and then you increase numHops. Wouldn't other nodes with the same number of hops not be expanded?

---

*[query/shortest.go, line 197 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcQ5Wh1mzg4d2YjuigC:-KcQ5Wh2TZcTImReSui7:bw2ue1u) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L197)):*
> ```Go
> 
> 		for _, childUID := range ul.Uids {
> 			if getPath(pc, childUID, to, path, cost+1, finalCost) {
> ```

Potentially exponential in number of children of subgraph? Why not item just store the parent node which is a (predicate, uid)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/570)*
<!-- Sent from Reviewable.io -->

gacharya255 commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[query/shortest.go, line 157 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcQ4ndR3p1So4MG7enV:-KcQBOgIl7kq7kfpai3I:b-c4fhca) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L157)):*
<details><summary><i>Previously, gacharya255 (Ganesh Acharya) wrote…</i></summary><blockquote>

I thought you have cost because in the future, it might be a variable. However, in that future case, the code may not work? Two hops may lead to lower cost than one hop, and you would need to update a node in the priority queue and move it around?
</blockquote></details>

As discussed, dijkstra would require updating of costs right at the time you pop the min node. There's no cost update here.

---

*[query/shortest.go, line 160 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcQ3zHpyhFcoxP_6ETd:-KcQBNG832JSVPPb1Mbe:bbstcb7) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L160)):*
<details><summary><i>Previously, gacharya255 (Ganesh Acharya) wrote…</i></summary><blockquote>

The first time you see a node at a fixed distance (num hops), you find its neighbors and then you increase numHops. Wouldn't other nodes with the same number of hops not be expanded?
</blockquote></details>

Ok i think i understand this now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/570)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 3 of 5 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks broke.

---

*[query/shortest.go, line 49 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR2sb6YbI8ORBtU3LR:-KcR2sb6YbI8ORBtU3LS:betauwn) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L49)):*
> ```Go
> }
> 
> func execNextLevel(ctx context.Context, start *SubGraph, next chan bool, res chan info, rch chan error) {
> ```

line too long. Call it `expandOut(..)`.

---

*[query/shortest.go, line 118 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR2MyjxjQdX3S8TY_f:-KcR2MykfheFrWDp2LF0:b-fx6392) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L118)):*
> ```Go
> 	var err error
> 	if sg.Params.Alias != "shortest" {
> 		rch <- nil
> ```

Should be error, right?

---

*[query/shortest.go, line 125 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR2Rlbi5O-HI--imVS:-KcR2Rlbi5O-HI--imVT:b-9pj3u0) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L125)):*
> ```Go
> 	heap.Init(&pq)
> 
> 	srcNode := &Item{sg.Params.From, 0, 0, 0}
> ```

Specify the variable names in the struct.

---

*[query/shortest.go, line 168 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR90TH0n5PQUb2dgPi:-KcR90TH0n5PQUb2dgPj:b-cdsfz6) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L168)):*
> ```Go
> 			}
> 
> 			for k, _ := range mp {
> ```

This shouldn't happen here. We shouldn't delete the neighbors from map. Also, instead of calling it mp, call it `neighbors`.

---

*[query/shortest.go, line 171 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR7xpbpQvJCsXkQVut:-KcR7xpcMdJTTpKvvSp7:b-4ajvx5) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L171)):*
> ```Go
> 				delete(mp, k)
> 			}
> 			for it := range res {
> ```

This should be a for loop with select over ctx.Done():

---

*[query/shortest.go, line 173 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR819XtWGxyZuYl3_w:-KcR819XtWGxyZuYl3_x:b-wiunb6) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L173)):*
> ```Go
> 			for it := range res {
> 				if it.to == 0 {
> 					break
> ```

break the for loop.

---

*[query/shortest.go, line 175 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcRAMxemfnEUTnfd6YD:-KcRAMxf_i3Vw-vKU4UK:bdrnjn6) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L175)):*
> ```Go
> 					break
> 				}
> 				if item.uid == it.from {
> ```

Keep things simple, push to map straight away. Then have the map be the source of information.

---

*[query/shortest.go, line 185 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcRA2bAS01hT_tTYgXr:-KcRA2bAS01hT_tTYgXs:b-v3u46) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L185)):*
> ```Go
> 				} else {
> 					//Put it in map.
> 					mp[it.from] = append(mp[it.from], it)
> ```

Add a comment that you shouldn't be getting a duplicate here, because the it edge is at least numHops edges out. There might be cycles though, which need to be taken care of -- best to take care of them before `expandOut`. Either that, or just do a map of map of struct{}.

Actually, when creating SrcUids, do a filter over DestUids with this map. In that case, just make expandOut a normal function, which this one calls and blocks on. Just pass this map directly.

---

*[query/shortest.go, line 191 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcRAscyiFWk0qkRLL2k:-KcRAsczm6jraN5gyB9U:btgfcd7) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L191)):*
> ```Go
> 		} else {
> 			// look at the store map.
> 			neigh := mp[item.uid]
> ```

delete the map here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/570)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: 4 of 5 files reviewed at latest revision, 14 unresolved discussions.

---

*[query/shortest.go, line 92 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcPwT6_zGWATDa7Dn0_:-KcR_Z2WG5JE6Z96xmj7:bmwtcq4) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L92)):*
<details><summary><i>Previously, gacharya255 (Ganesh Acharya) wrote…</i></summary><blockquote>

Checking my understanding: The expansion is only along start.Children which is why you need to call ProcessGraph again and again here.
</blockquote></details>

That's correct.

---

*[query/shortest.go, line 157 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcQ4ndR3p1So4MG7enV:-KcRQBfrLS59xAC2BH_Q:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L157)):*
<details><summary><i>Previously, gacharya255 (Ganesh Acharya) wrote…</i></summary><blockquote>

As discussed, dijkstra would require updating of costs right at the time you pop the min node. There's no cost update here.
</blockquote></details>

Done.

---

*[query/shortest.go, line 160 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcQ3zHpyhFcoxP_6ETd:-KcRZZTFpLJd2MNrCpyY:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L160)):*
<details><summary><i>Previously, gacharya255 (Ganesh Acharya) wrote…</i></summary><blockquote>

Ok i think i understand this now.
</blockquote></details>

Done.

---

*[query/shortest.go, line 197 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcQ5Wh1mzg4d2YjuigC:-KcRZ_NQK1RrEkh6W-S7:b-pc862f) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L197)):*
<details><summary><i>Previously, gacharya255 (Ganesh Acharya) wrote…</i></summary><blockquote>

Potentially exponential in number of children of subgraph? Why not item just store the parent node which is a (predicate, uid)?
</blockquote></details>

Yeah, that would be another way to get the path. We'd use an extra map.

---

*[query/shortest.go, line 49 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR2sb6YbI8ORBtU3LR:-KcR_XQ4vV8t3LTduVcQ:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L49)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

line too long. Call it `expandOut(..)`.
</blockquote></details>

Done.

---

*[query/shortest.go, line 118 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR2MyjxjQdX3S8TY_f:-KcRZQ28u9FKdqPN2L0e:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L118)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Should be error, right?
</blockquote></details>

Done.

---

*[query/shortest.go, line 125 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR2Rlbi5O-HI--imVS:-KcR_C2ERmRmP3l-rW2Q:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L125)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Specify the variable names in the struct.
</blockquote></details>

Done.

---

*[query/shortest.go, line 168 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR90TH0n5PQUb2dgPi:-KcRZDZYk4QQqbUkqjVq:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L168)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This shouldn't happen here. We shouldn't delete the neighbors from map. Also, instead of calling it mp, call it `neighbors`.
</blockquote></details>

Done.

---

*[query/shortest.go, line 171 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR7xpbpQvJCsXkQVut:-KcRZC_xl1gYnpRHn1v3:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L171)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This should be a for loop with select over ctx.Done():
</blockquote></details>

Done.

---

*[query/shortest.go, line 173 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR819XtWGxyZuYl3_w:-KcRZCDlggPhHgcFhYK8:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L173)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

break the for loop.
</blockquote></details>

Done.

---

*[query/shortest.go, line 175 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcRAMxemfnEUTnfd6YD:-KcRZAIqFDooPjKYLAVf:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L175)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Keep things simple, push to map straight away. Then have the map be the source of information.
</blockquote></details>

Done.

---

*[query/shortest.go, line 185 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcRA2bAS01hT_tTYgXr:-KcRZUx2LoGgLQndbvQ_:b-9798w9) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L185)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Add a comment that you shouldn't be getting a duplicate here, because the it edge is at least numHops edges out. There might be cycles though, which need to be taken care of -- best to take care of them before `expandOut`. Either that, or just do a map of map of struct{}.

Actually, when creating SrcUids, do a filter over DestUids with this map. In that case, just make expandOut a normal function, which this one calls and blocks on. Just pass this map directly.
</blockquote></details>

Changed the logic a bit. Taking care of loops.

---

*[query/shortest.go, line 191 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcRAscyiFWk0qkRLL2k:-KcRZZ32Ur8EaQ2CyMRM:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L191)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

delete the map here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/570)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Review status: 4 of 5 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[query/shortest.go, line 197 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcQ5Wh1mzg4d2YjuigC:-KcWZHbz5XRuugkNJCdU:b-8plhe1) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L197)):*
<details><summary><i>Previously, ashwin95r (Ashwin Ramesh) wrote…</i></summary><blockquote>

Yeah, that would be another way to get the path. We'd use an extra map.
</blockquote></details>

Better to iterate than to store a map per node in the graph.

---

*[query/shortest.go, line 109 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWUh8bDOm6fD6ZUyYx:-KcWUh8bDOm6fD6ZUyYy:blvi6g1) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L109)):*
> ```Go
> 			return
> 		}
> ```

vertical space

---

*[query/shortest.go, line 140 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWVrVgj0Qrr5vyAD4H:-KcWVrVgj0Qrr5vyAD4I:bcvtgmh) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L140)):*
> ```Go
> }
> 
> func ShortestPath(ctx context.Context, sg *SubGraph, rch chan error) {
> ```

Same thing for here. I think error handling and returns become simpler if you just have the function return an error; instead of pushing it to a channel.

Having a channel makes sense if this same channel is being used by many different goroutines. Might be the case here, I'm not sure.

---

*[query/shortest.go, line 156 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWWBIwItUaXWc1tXki:-KcWWBIwItUaXWc1tXkj:b-wjtrdc) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L156)):*
> ```Go
> 		index: 0,
> 	}
> ```

Avoid unnecessary vertical spaces after a line with only a closing bracket.

---

*[query/shortest.go, line 163 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWWOgg7-FglyUorYA7:-KcWWOgg7-FglyUorYA8:bknun6e) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L163)):*
> ```Go
> 
> 	next := make(chan bool, 2)
> 	rch1 := make(chan error, 2)
> ```

Call it expandErr. In general, variable names ending with a digit are bad names.

---

*[query/shortest.go, line 165 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWVPjFkL_4NQ6elUSr:-KcWVPjFkL_4NQ6elUSs:bbr6mkg) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L165)):*
> ```Go
> 	rch1 := make(chan error, 2)
> 	mp := make(map[uint64]map[uint64]float64)
> 	go sg.expandOut(ctx, mp, next, rch1)
> ```

Better to have sg.expandOut return an error, and push it to channel here. Would make the error handling code simpler for expandOut as well.

---

*[query/shortest.go, line 174 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWaOK1jehqgnSWwyoV:-KcWaOK1jehqgnSWwyoW:b71n6i0) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L174)):*
> ```Go
> 
> 	// For now, lets allow a maximum of 10 hops.
> 	for pq.Len() > 0 && numHops < 10 {
> ```

Instead of doing this, we can do this based on the map size. So, we don't end up taking too much memory.

---

*[query/shortest.go, line 177 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWVcMsheWlrMGAzEye:-KcWVcMsheWlrMGAzEyf:bof287h) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L177)):*
> ```Go
> 		item := heap.Pop(&pq).(*Item)
> 		/*
> 			if item.hop > maxLoopLevel {
> ```

Don't you need this?

---

*[query/shortest.go, line 195 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWVhIpJy-ABWklXnR0:-KcWVhIpJy-ABWklXnR1:b5w9ss2) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L195)):*
> ```Go
> 			case err = <-rch1:
> 				if err != nil {
> 					if err.Error() == "Stop Expansion" {
> ```

Not idiomatic Go. You should declare a const error, and use that instead.

---

*[query/shortest.go, line 211 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWX1W7HbHF1bauGl_y:-KcWX1W7HbHF1bauGl_z:b-3abjnv) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L211)):*
> ```Go
> 			numHops++
> 		}
> 		neigh := mp[item.uid]
> ```

What does `neigh` mean dude?

---

*[query/shortest.go, line 214 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWZ9Wh0ZK1rc6zU2el:-KcWZ9Wh0ZK1rc6zU2em:b-lr98pg) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L214)):*
> ```Go
> 		for toUid, cost := range neigh {
> 			if d, ok := dist[toUid]; !ok || d > item.cost+cost {
> 				node := &Item{toUid, item.cost + cost, item.hop + 1, 0}
> ```

Use the fields as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/570)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Review status: 3 of 5 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[query/query.go, line 582 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcXGWLfXl7pk_WIDJu5:-KcXGWLfXl7pk_WIDJu6:b-pvkzr9) ([raw file](https://github.com/dgraph-io/dgraph/blob/034a418756740379b7b86a0c9865cb958f06ca1d/query/query.go#L582)):*
> ```Go
> 	for i := 0; i < len(res.Query); i++ {
> 		gq := res.Query[i]
> 		if gq == nil || (len(gq.UID) == 0 && gq.Func == nil && len(gq.NeedsVar) == 0 && gq.Alias != "shortest") {
> ```

Long line.

---

*[query/query.go, line 709 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcXGrb_tcpa5WwsY0ne:-KcXGrbaqWjlGa3L62lY:bepuk3d) ([raw file](https://github.com/dgraph-io/dgraph/blob/034a418756740379b7b86a0c9865cb958f06ca1d/query/query.go#L709)):*
> ```Go
> func populateVarMap(sg *SubGraph, doneVars map[string]*task.List, isCascade bool) {
> 	out := sg.DestUIDs.Uids[:0]
> 	if sg.Params.Alias == "shortest" {
> ```

If someone assigns a different alias for shortest, this won't work? Or, do you enforce that one can't assign an alias to shortest?

---

*[query/shortest.go, line 49 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR2sb6YbI8ORBtU3LR:-KcXHINBCxdLF8zBvcMa:b8pny6p) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L49)):*
<details><summary><i>Previously, ashwin95r (Ashwin Ramesh) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

still too long. Also remove rch chan error.

---

*[query/shortest.go, line 174 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWaOK1jehqgnSWwyoV:-KcXKZnjny7BB4lGeCcr:buqfqpr) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L174)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Instead of doing this, we can do this based on the map size. So, we don't end up taking too much memory.
</blockquote></details>

Map size here is really the sum of lengths of all maps of maps, not just one map.

---

*[query/shortest.go, line 92 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcXJHVoEDk2xzn3a1bt:-KcXJHVoEDk2xzn3a1bu:b-r8zmfl) ([raw file](https://github.com/dgraph-io/dgraph/blob/034a418756740379b7b86a0c9865cb958f06ca1d/query/shortest.go#L92)):*
> ```Go
> 					}
> 					if _, ok := mp[fromUID][toUid]; !ok {
> 						isNew = true
> ```

This is strange logic. Even if one node is repeated, you'd stop expanding all nodes?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/570)*
<!-- Sent from Reviewable.io -->

gacharya255 commented:
Review status: 3 of 5 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[query/shortest.go, line 197 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcQ5Wh1mzg4d2YjuigC:-KcYR84cf8Rgip3_8Q94:b6aw560) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L197)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Better to iterate than to store a map per node in the graph.
</blockquote></details>

1) When the cost is not uniform, you might have trouble finding this path.
2) You seem to be doing a DFS / brute force search here which defeats the purpose of dijkstra and is not just slow, but exponentially slower.
3) You can store this extra information together with the cost. It is not really an additional map.

---

*[query/shortest.go, line 200 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcYQBB0giXenLbxEGxb:-KcYQBB0giXenLbxEGxc:byh9ye6) ([raw file](https://github.com/dgraph-io/dgraph/blob/034a418756740379b7b86a0c9865cb958f06ca1d/query/shortest.go#L200)):*
> ```Go
> 					hop:  item.hop + 1,
> 				}
> 				heap.Push(&pq, node) // Add a node with lesser cost in the queue.
> ```

I thought you will use Fix or at least do a Remove. Otherwise,

1) heap might have O(n^2) elements instead of O(n) where n is number of vertices.
2) you might revisit the same node and waste computation, again O(n^2).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/570)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: 3 of 5 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[query/query.go, line 582 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcXGWLfXl7pk_WIDJu5:-Kc_XLvGhdTRs1leiXPg:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/034a418756740379b7b86a0c9865cb958f06ca1d/query/query.go#L582)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Long line.
</blockquote></details>

Done.

---

*[query/query.go, line 709 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcXGrb_tcpa5WwsY0ne:-KcaBJn-lNiVSD55B0Wb:b-bxusd7) ([raw file](https://github.com/dgraph-io/dgraph/blob/034a418756740379b7b86a0c9865cb958f06ca1d/query/query.go#L709)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

If someone assigns a different alias for shortest, this won't work? Or, do you enforce that one can't assign an alias to shortest?
</blockquote></details>

shortest is at the root. We can't have aliases at root. But if someone names other predicates alias as shortest that might cause a problem. We need to finalise a format for this

---

*[query/shortest.go, line 16 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcPune1NDMOTSf5CCFP:-KcWo6oxVuH6U6mfokMu:bdzatzz) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L16)):*
<details><summary><i>Previously, gacharya255 (Ganesh Acharya) wrote…</i></summary><blockquote>

index tells you when it is added but don't think you really need that?
</blockquote></details>

Right, we don't need it for now, since we add a new node instead of update.

---

*[query/shortest.go, line 197 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcQ5Wh1mzg4d2YjuigC:-Kc_kX3jb8ZuEyg3mu0B:bepqvej) ([raw file](https://github.com/dgraph-io/dgraph/blob/50602fac8441db3db8422af382b6f80c1ed505f1/query/shortest.go#L197)):*
<details><summary><i>Previously, gacharya255 (Ganesh Acharya) wrote…</i></summary><blockquote>

1) When the cost is not uniform, you might have trouble finding this path.
2) You seem to be doing a DFS / brute force search here which defeats the purpose of dijkstra and is not just slow, but exponentially slower.
3) You can store this extra information together with the cost. It is not really an additional map.
</blockquote></details>

Done. Agree.

---

*[query/shortest.go, line 49 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcR2sb6YbI8ORBtU3LR:-Kc_XYeEbjTL0ZNxgC5r:b-qu5awf) ([raw file](https://github.com/dgraph-io/dgraph/blob/87e1fb8faa234a92ae0f7fdb01bc280deb18ad4a/query/shortest.go#L49)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

still too long. Also remove rch chan error.
</blockquote></details>

As discussed we need rch to keep the state maintenance clean.

---

*[query/shortest.go, line 109 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWUh8bDOm6fD6ZUyYx:-KcWkpyBRLIPawpATk8Q:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L109)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

vertical space
</blockquote></details>

Done.

---

*[query/shortest.go, line 140 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWVrVgj0Qrr5vyAD4H:-KcWmeB_9Q7M6kMLvPxa:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L140)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Same thing for here. I think error handling and returns become simpler if you just have the function return an error; instead of pushing it to a channel.

Having a channel makes sense if this same channel is being used by many different goroutines. Might be the case here, I'm not sure.
</blockquote></details>

Done.

---

*[query/shortest.go, line 156 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWWBIwItUaXWc1tXki:-KcWmsaAPi3DeizJL7Gd:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L156)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Avoid unnecessary vertical spaces after a line with only a closing bracket.
</blockquote></details>

Done.

---

*[query/shortest.go, line 163 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWWOgg7-FglyUorYA7:-KcWn-p7SuC3GZU5O2I7:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L163)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Call it expandErr. In general, variable names ending with a digit are bad names.
</blockquote></details>

Done.

---

*[query/shortest.go, line 165 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWVPjFkL_4NQ6elUSr:-Kc_kaAi42f-MC_xhzBN:b-wuqjnv) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L165)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Better to have sg.expandOut return an error, and push it to channel here. Would make the error handling code simpler for expandOut as well.
</blockquote></details>

As discussed, we just returning wont work.

---

*[query/shortest.go, line 174 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWaOK1jehqgnSWwyoV:-KcWnmqo-Uzihnne3qkZ:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L174)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Instead of doing this, we can do this based on the map size. So, we don't end up taking too much memory.
</blockquote></details>

Done.

---

*[query/shortest.go, line 177 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWVcMsheWlrMGAzEye:-KcWnA4V8vPiqu8s7wwl:b94uaao) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L177)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't you need this?
</blockquote></details>

Nope, stopExpansion should do this now.

---

*[query/shortest.go, line 195 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWVhIpJy-ABWklXnR0:-Kc_VwWDt_zhzkwIavms:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L195)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Not idiomatic Go. You should declare a const error, and use that instead.
</blockquote></details>

Done.

---

*[query/shortest.go, line 211 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWX1W7HbHF1bauGl_y:-KcWnpSG8xPWIn-G4neO:b-wzykfq) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L211)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

What does `neigh` mean dude?
</blockquote></details>

neighbours. Changed it.

---

*[query/shortest.go, line 214 at r3](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcWZ9Wh0ZK1rc6zU2el:-KcWnXRDVECC2eqyU9Zj:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/05a21d171780dc747ca2e6b34a94fc797a1759d5/query/shortest.go#L214)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Use the fields as well.
</blockquote></details>

Done.

---

*[query/shortest.go, line 92 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcXJHVoEDk2xzn3a1bt:-Kc_XrRKUXHPAwz9mKtf:b-jjq8xy) ([raw file](https://github.com/dgraph-io/dgraph/blob/034a418756740379b7b86a0c9865cb958f06ca1d/query/shortest.go#L92)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This is strange logic. Even if one node is repeated, you'd stop expanding all nodes?
</blockquote></details>

Moved it to filter step, so we anyways don't need this.

---

*[query/shortest.go, line 200 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcYQBB0giXenLbxEGxb:-Kca8ICWQj6sjLkq5SUM:b-aj7bmz) ([raw file](https://github.com/dgraph-io/dgraph/blob/034a418756740379b7b86a0c9865cb958f06ca1d/query/shortest.go#L200)):*
<details><summary><i>Previously, gacharya255 (Ganesh Acharya) wrote…</i></summary><blockquote>

I thought you will use Fix or at least do a Remove. Otherwise,

1) heap might have O(n^2) elements instead of O(n) where n is number of vertices.
2) you might revisit the same node and waste computation, again O(n^2).
</blockquote></details>

That's right. Working on improving this by keeping track of the nodes position in the priority queue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/570)*
<!-- Sent from Reviewable.io -->

gacharya255 commented:
Logic seems correct. Just some minor comments now. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: 3 of 5 files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[query/shortest.go, line 47 at r5](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcaJp0Yxsv9oHJQPZgv:-KcaJp0ZPhBRA67Md1Cy:b-s2vdau) ([raw file](https://github.com/dgraph-io/dgraph/blob/c33bc7e01dcffbbe4addcfe1697dfc22aa08b9cd/query/shortest.go#L47)):*
> ```Go
> }
> 
> type minInfo struct {
> ```

Maybe rename minInfo to something like nodeInfo? And maybe a comment saying that we maintain a map from UID to minInfo/nodeInfo for Dijkstra.

---

*[query/shortest.go, line 50 at r5](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcaK5rWCnN29LvI1RH4:-KcaK5rWCnN29LvI1RH5:b5u3rp6) ([raw file](https://github.com/dgraph-io/dgraph/blob/c33bc7e01dcffbbe4addcfe1697dfc22aa08b9cd/query/shortest.go#L50)):*
> ```Go
> 	cost   float64
> 	parent uint64
> 	node   *Item
> ```

To help us remember, do you mind adding a comment here saying: "node points to item in priority queue so that we can update cost."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-:-KcaQmWzGuCaeLqmTIXe:b-a036k9)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: 3 of 5 files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[query/shortest.go, line 47 at r5](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcaJp0Yxsv9oHJQPZgv:-KcaSxTeiX7PdUhQvBtK:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c33bc7e01dcffbbe4addcfe1697dfc22aa08b9cd/query/shortest.go#L47)):*
<details><summary><i>Previously, gacharya255 (Ganesh Acharya) wrote…</i></summary><blockquote>

Maybe rename minInfo to something like nodeInfo? And maybe a comment saying that we maintain a map from UID to minInfo/nodeInfo for Dijkstra.
</blockquote></details>

Done.

---

*[query/shortest.go, line 50 at r5](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcaK5rWCnN29LvI1RH4:-KcaSx0wLZQVC0-gKBCU:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c33bc7e01dcffbbe4addcfe1697dfc22aa08b9cd/query/shortest.go#L50)):*
<details><summary><i>Previously, gacharya255 (Ganesh Acharya) wrote…</i></summary><blockquote>

To help us remember, do you mind adding a comment here saying: "node points to item in priority queue so that we can update cost."
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/570)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 26 unresolved discussions, some commit checks failed.

---

*[query/query.go, line 769 at r6](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcqdOO5WxAAkW_Xtv2V:-KcqdOO5WxAAkW_Xtv2W:b-n9q2vc) ([raw file](https://github.com/dgraph-io/dgraph/blob/306063f5687c7b563e6b3fb52095c91d563bec0b/query/query.go#L769)):*
> ```Go
> // TODO(Ashwin): Benchmark this function.
> func populateVarMap(sg *SubGraph, doneVars map[string]*task.List, isCascade bool) {
> 	o := new(task.List)
> ```

var o task.List

---

*[query/query.go, line 770 at r6](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcqdPhYbIaI0Qai3Fqd:-KcqdPhYbIaI0Qai3Fqe:b-48b3za) ([raw file](https://github.com/dgraph-io/dgraph/blob/306063f5687c7b563e6b3fb52095c91d563bec0b/query/query.go#L770)):*
> ```Go
> func populateVarMap(sg *SubGraph, doneVars map[string]*task.List, isCascade bool) {
> 	o := new(task.List)
> 	out := algo.NewWriteIterator(o)
> ```

out := algo.NewWriteIterator(&o)

---

*[query/shortest.go, line 61 at r6](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-Kcqe66kfRNL9kdrR-Vo:-Kcqe66kfRNL9kdrR-Vp:b-qj7dcm) ([raw file](https://github.com/dgraph-io/dgraph/blob/306063f5687c7b563e6b3fb52095c91d563bec0b/query/shortest.go#L61)):*
> ```Go
> 	var exec []*SubGraph
> 	var err error
> 	in := new(task.List)
> ```

var in task.List

---

*[query/shortest.go, line 250 at r6](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcqfZqGs0xhy0IjQREB:-KcqfZqGs0xhy0IjQREC:bjkgx6m) ([raw file](https://github.com/dgraph-io/dgraph/blob/306063f5687c7b563e6b3fb52095c91d563bec0b/query/shortest.go#L250)):*
> ```Go
> 			neighbours := adjacencyMap[item.uid]
> 			for toUid, cost := range neighbours {
> 				if d, ok := dist[toUid]; !ok || d.cost > item.cost+cost {
> ```

Too much indentation. Handle the other cond first.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-:-KcqgKa_f7fBrx2UE16Y:bnfp4nl)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: all files reviewed at latest revision, 26 unresolved discussions, some commit checks failed.

---

*[query/query.go, line 769 at r6](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcqdOO5WxAAkW_Xtv2V:-Kcqh_TXy11XYLJx4n0k:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/306063f5687c7b563e6b3fb52095c91d563bec0b/query/query.go#L769)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

var o task.List
</blockquote></details>

Done.

---

*[query/query.go, line 770 at r6](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcqdPhYbIaI0Qai3Fqd:-Kcqh_twqzH1P-FbMK-g:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/306063f5687c7b563e6b3fb52095c91d563bec0b/query/query.go#L770)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

out := algo.NewWriteIterator(&o)
</blockquote></details>

Done.

---

*[query/shortest.go, line 200 at r4](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcYQBB0giXenLbxEGxb:-Kcqh7yS0Lx6PBTw9W1-:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/034a418756740379b7b86a0c9865cb958f06ca1d/query/shortest.go#L200)):*
<details><summary><i>Previously, ashwin95r (Ashwin Ramesh) wrote…</i></summary><blockquote>

That's right. Working on improving this by keeping track of the nodes position in the priority queue.
</blockquote></details>

Done.

---

*[query/shortest.go, line 61 at r6](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-Kcqe66kfRNL9kdrR-Vo:-KcqhxNxd1Q4hSxxE9DY:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/306063f5687c7b563e6b3fb52095c91d563bec0b/query/shortest.go#L61)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

var in task.List
</blockquote></details>

Done.

---

*[query/shortest.go, line 250 at r6](https://reviewable.io:443/reviews/dgraph-io/dgraph/570#-KcqfZqGs0xhy0IjQREB:-KcqiaCiwrTRxdwABzBx:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/306063f5687c7b563e6b3fb52095c91d563bec0b/query/shortest.go#L250)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Too much indentation. Handle the other cond first.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/570)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Merged.
